R/unlockREDCap.R

```diff
@@ -15,15 +14,120 @@
 # You should have received a copy of the GNU General Public License
 # along with this program.  If not, see <http://www.gnu.org/licenses/>.
 
+  ###########################################################################
+ ## Connection function
+##
+.connectAndCheck <- function(key, url, ...)
+{
+  tryCatch(
+    { 
+      conn <- redcapConnection(key, url=url, ...)
+      conn$metadata() # Test connection by reading metadata into cache
+      conn
+    },
+    error = function(e)
+    {
+      if(grepl("Could not resolve host",     e) ||
+         grepl("Couldn't connect to server", e))
+        stop("Unable to connect to url '",url,"'. ", e$message)
+        
+      if(grepl("403", e)) return(NULL)
+      
+      stop(e)
+    }
+  )
+}
 
- #############################################################################
-## Check if key is in package environment, aka memory
-.key_saved <- function(envir, key)
+  #############################################################################
+ ## unlock via YAML override if it exists
+##
+.unlockYamlOverride <- function(connections, url, ...)
 {
-  exists(key, envir=envir, inherits=FALSE) &&
-    !is.null(envir[[key]])                   &&
-    !is.na(envir[[key]])                     &&
-    !envir[[key]]==''
+  config_file <- file.path("..", paste0(basename(getwd()),".yml"))
+  
+  if(!file.exists(config_file)) return(list())
+  
+  config <- read_yaml(config_file)
+  if(is.null(config$redcapAPI)) stop(paste0("Config file '",config_file,"' does not contain required 'redcapAPI' entry"))
+  config <- config$redcapAPI
+  if(is.null(config$keys))      stop(paste0("Config file '",config_file,"' does not contain required 'keys' entry under the 'redcapAPI' entry"))
+  keys   <- config$keys
+  if(!is.null(config$args$url))  url <- config$args$url # Override from yml if available
+  config$args$url <- NULL
+  args   <- c(config$args, url = url, list(...))
+    
+  dest <- lapply(seq_along(connections), function(i) 
+  {
+    args$key  <- keys[connections[i]]
+    
+    if(is.null(args$key)) stop(paste0("Config file '", config_file, "' does not have API_KEY for '",connections[i],"' under redcapAPI: keys: specified."))
+    
+    do.call(.connectAndCheck, args)
+  })
+  names(dest) <- if(is.null(names(connections))) connections else names(connections)
+  
+  return(dest)
+}
+ 
+  #############################################################################
+ ## unlock keyring
+##
+.unlockKeyring <- function(keyring, passwordFUN)
+{
+  state <- keyring_list()
```

Comment:
```suggestion
  state <- keyring::keyring_list()
```

Reply:
I changed this to the non-namespaced version to work with the stubbed tests.